docs/proposals/12-oidc-webhook-authenticator.md

```diff
@@ -0,0 +1,305 @@
+
+![alt text](assets/registration.svg "Authentication with OIDC webhook")
+
+### End-user authentication via new OpenIDConnect IDP
+
+When a user presents an `id_token` obtained from a OpenID Connect the flow looks like this:
+
+1. The user authenticates in Custom IDP.
+1. `id_token` is obtained from Custom IDP.
+1. The user uses `id_token` to perform an API call to Kube APIServer.
+1. As the `id_token` is not matched by any build-in or configured authenticators in the Kube APIServer, it is send to OWA for validation.
+
+    ```json
+    {
+      "TokenReview": {
+        "kind": "TokenReview",
+        "apiVersion": "authentication.k8s.io/v1beta1",
+        "spec": {
+          "token": "ddeewfwef..."
+        }
+      }
+    }
+    ```
+
+1. OWA uses `TokenReview` to authenticate the calling API server (the Kube APIServer for delegation of authentication and authorization is different from the calling API server)
+
+    ```json
+    {
+      "TokenReview": {
+        "kind": "TokenReview",
+        "apiVersion": "authentication.k8s.io/v1beta1",
+        "spec": {
+          "token": "api-server-token..."
+        }
+      }
+    }
+    ```
+
+1. After the Authentication API server returns the identity of callee API server:
+
+    ```json
+    {
+        "apiVersion": "authentication.k8s.io/v1",
+        "kind": "TokenReview",
+        "metadata": {
+            "creationTimestamp": null
+        },
+        "spec": {
+            "token": "eyJhbGciOiJSUzI1NiIsImtpZCI6InJocEdLTXZlYjV1OE5heD..."
+        },
+        "status": {
+            "authenticated": true,
+            "user": {
+                "groups": [
+                    "system:serviceaccounts",
+                    "system:serviceaccounts:shoot--abcd",
+                    "system:authenticated"
+                ],
+                "uid": "14db103e-88bb-4fb3-8efd-ca9bec91c7bf",
+                "username": "system:serviceaccount:shoot--abcd:kube-apiserver"
+            }
+        }
+    }
+    ```
+
+    OWA makes a `SubjectAccessReview` call to the Authorization API server:
+
+    ```json
+    {
+      "apiVersion": "authorization.k8s.io/v1",
+      "kind": "SubjectAccessReview",
+      "spec": {
+        "groups": [
+          "system:serviceaccounts",
+          "system:serviceaccounts:shoot--abcd",
+          "system:authenticated"
+        ],
+        "nonResourceAttributes": {
+          "path": "/validate-token",
+          "verb": "post"
+        },
+        "user": "system:serviceaccount:shoot--abcd:kube-apiserver"
+      },
+      "status": {
+        "allowed": true,
+        "reason": "RBAC: allowed by RoleBinding \"kube-apiserver\" of ClusterRole \"kube-apiserver\" to ServiceAccount \"system:serviceaccount:shoot--abcd:kube-apiserver\""
+      }
+    }
+    ```
+
+1. OWA then iterates over all registered `OpenIDConnect` and tries to validate the token.
```

Comment:
```suggestion
1. OWA then iterates over all registered `OpenIDConnect` providers and tries to validate the token.
```

Reply:
Maybe also something to think about in advance: 
- do we enforce a timeout (configurable?)  if an OIDC provider is unresponsive? 
- how to make sure that the original request does not time out when looping over multiple OIDC providers?

Reply:
There is no network traffic in this case - jwt / oidc authenticator is constructed with `jwks`, `clientid`, `issuer` and so. The validation flow in this case is:

- parse token string into `jose.JSONWebSignature`
- iterate over all token authenticators (in-memory) until one of them matches the signature

Reply:
ok true, its offline JWT token verification. Thanks

Reply:
Actually this is not entirely correct - if the token is verified to be valid for an issuer, client ID .. etc.. it might have [distributed claims](https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims) in which case they might be fetched and optionally used for further verification if `requiredClaims` is set.

Reply:
I think I have overlooked this. 
So that means my original concerns regarding timeout do apply? We do not have to include it in the GEP, just wanted to mention the point.

Reply:
ping